File: testapp/src/main/java/com/appliedrec/mrtd_reader_app/MainActivity.java
package com.appliedrec.mrtd_reader_app;

import android.content.Intent;
import android.os.Bundle;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.preference.PreferenceManager;

import com.appliedrec.mrtd_reader_sample.R;
import com.appliedrec.mrtd_reader_sample.databinding.ActivityMainBinding;
import com.appliedrec.mrtdreader.BACSpec;
import com.appliedrec.mrtdreader.MRTDScanResult;
import com.appliedrec.mrtdreader.MRTDScanSession;
import com.appliedrec.mrtdreader.MRTDScanSessionListener;

public class MainActivity extends AppCompatActivity implements MRTDScanSessionListener, BACEntryFragment.Listener {

    private BACSpecModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ActivityMainBinding viewBinding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(viewBinding.getRoot());
        viewModel = new ViewModelProvider(this).get(BACSpecModel.class);
        viewModel.setSharedPreferences(PreferenceManager.getDefaultSharedPreferences(this));
        if (savedInstanceState == null) {
            getSupportFragmentManager().beginTransaction().add(viewBinding.container.getId(), BACEntryFragment.newInstance(viewModel.getBACSpec().getValue())).commit();
        }
    }

    //region MRTD scan listener

    @Override
    public void onMRTDScanSucceeded(BACSpec bacSpec, MRTDScanResult result) {
        Intent intent = new Intent(this, CaptureResultActivity.class);
        intent.putExtra(CaptureResultActivity.EXTRA_MRTD_SCAN_RESULT, result);
        startActivity(intent);
    }

    @Override
    public void onMRTDScanFailed(BACSpec bacSpec, Throwable throwable) {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(R.string.failed_to_read_travel_document);
        builder.setMessage(throwable.toString());
        builder.setNeutralButton(android.R.string.ok, null);
        builder.create().show();
    }

    @Override
    public void onMRTDScanCancelled(BACSpec bacSpec) {

    }

    //endregion

    @Override
    public void onRequestCapture(BACSpec bacSpec) {
        MRTDScanSession session = new MRTDScanSession(this, bacSpec);
        session.setListener(this);
        session.start();
    }
}

File: testapp/src/main/java/com/appliedrec/mrtd_reader_app/MainActivity.kt
package com.appliedrec.mrtd_reader_app

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.preference.PreferenceManager
import com.appliedrec.mrtd_reader_app.BACSpecModel
import com.appliedrec.mrtd_reader_app.databinding.ActivityMainBinding
import com.appliedrec.mrtdreader.BACSpec
import com.appliedrec.mrtdreader.MRTDReaderActivityResultContract
import com.appliedrec.mrtdreader.MRTDScanResult
import com.appliedrec.mrtdreader.MRTDScanSettings
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class MainActivity : AppCompatActivity(), BACEntryFragment.Listener {
    private var viewModel: BACSpecModel? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val viewBinding = ActivityMainBinding.inflate(
            layoutInflater
        )
        setContentView(viewBinding.root)
        viewModel = ViewModelProvider(this).get(BACSpecModel::class.java)
        viewModel!!.setSharedPreferences(PreferenceManager.getDefaultSharedPreferences(this))
        if (savedInstanceState == null) {
            supportFragmentManager.beginTransaction().add(
                viewBinding.container.id, BACEntryFragment.newInstance(
                    viewModel!!.liveData.value
                )
            ).commit()
        }
    }

    private fun onMRTDScanSucceeded(result: MRTDScanResult.Success) {
        lifecycleScope.launch(Dispatchers.IO) {
            try {
                val uri = ResultFileHelper.saveScanResult(result)
                withContext(Dispatchers.Main) {
                    val intent = Intent(this@MainActivity, CaptureResultActivity::class.java)
                    intent.data = uri
                    startActivity(intent)
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    val builder = AlertDialog.Builder(this@MainActivity)
                    builder.setTitle(R.string.failed_to_save_scan_result)
                    builder.setMessage(e.toString())
                    builder.setNeutralButton(android.R.string.ok, null)
                    builder.create().show()
                }
            }
        }
    }

    private fun onMRTDScanFailed(throwable: Throwable) {
        val builder = AlertDialog.Builder(this)
        builder.setTitle(R.string.failed_to_read_travel_document)
        builder.setMessage(throwable.toString())
        builder.setNeutralButton(android.R.string.ok, null)
        builder.create().show()
    }

    private val launcher = registerForActivityResult(MRTDReaderActivityResultContract()) { result ->
        when (result) {
            is MRTDScanResult.Success -> onMRTDScanSucceeded(result)
            is MRTDScanResult.Failure -> onMRTDScanFailed(result.error)
            else -> {}
        }
    }

    override fun onRequestCapture(bacSpec: BACSpec) {
        launcher.launch(MRTDScanSettings(bacSpec))
    }
}